### ISSUE TITLE
GL3510

### ISSUE THREAD

**Issue by NothNoth** [c0d3z3r0/u2glhub#1]
Hi,

This is not really an issue, more a discussion (I was no able to find your email) :)

I'm working on the GL3510 chip which is "quite close" to the GL3523/GL3525.
There's no real documentation (except the datasheet) on the GL352x series, but at least one may find an ISP tool. When it comes to the GL3510 things get worse, there's no associated ISP tool available.

So, just like you did, I've partially reversed some of those ISP tools and had some successes with some commands.

Currently, I'm able to extract the Firmware version (1622) and.. extract "something" from the SRAM.
[XRAM_hex.txt](https://github.com/user-attachments/files/15965242/XRAM_hex.txt)
The SRAM dump is a bit strange, there's a repeated data block a 0x1000 and then 0x8000 at every 0x200 interval.

The Firmware dump fails, it returns a buffer filled with 0xFF.

I'm still trying to go deeper (with very few information..) but if you have some insights, I would be interested!

Cheers

 Ben

**Comment by c0d3z3r0** [comment]
Hi there!

> I was no able to find your email

Try harder :P It's possible ;)

> There's no real documentation (except the datasheet) on the GL352x series, but at least one may find an ISP tool. When it comes to the GL3510 things get worse, there's no associated ISP tool available.

I found out that the ISP tool is universal and support more than just one hub model. However, the leaked versions are pretty old. (AFAIK 4.2.0.0 is most recent, but not leaked). At least v2.7.4.11 supports GL3510.

https://station-drivers.com/index.php/en/component/remository/func-startdown/4173/lang,en-gb/

> The Firmware dump fails, it returns a buffer filled with 0xFF.

What device exactly do you have? Is there an external SPI flash? The GL3xxx hubs have a mask ROM (r/o) with basic hub firmware. Genesys Logic says that this is only for boot-up, but it still has basic hub functionality. A SPI flash is required for firmware and (simulated) eeprom settings but it's likely that some cheap hubs just ignore that requirement.

I was not able to dump the mask ROM on any hub so far.

> The SRAM dump is a bit strange, there's a repeated data block a 0x1000 and then 0x8000 at every 0x200 interval.

Yeah, I've seen that as well on GL352x:

0x1F00 maybe SRAM?
0x2000 maybe SRAM?
0x4000 mapped SPI flash

SPI flash is mapped at 0x4000-0xFFFF on GL352x. When the SPI flash is empty, I see "random" stuff there as well. I guess your board either doesn't have a SPI flash or it's empty.

Looking at your dump, 0x8000 on GL3510 seems to be equivalent to 0x4000 on GL352x.

**Comment by NothNoth** [comment]
You were right, I guess I found your email ;)
I sent you a message

**Comment by c0d3z3r0** [comment]
Nope, didn't get a message :D

**Comment by NothNoth** [comment]
Oh? I did found something on dmx.de, maybe this is an old one :)

**Comment by c0d3z3r0** [comment]
yeah, the gmx one is veeeery old and doesn't exist anymore. use foss@mniewoehner.de
